### PR TITLE
Extract foreign key action from `information_schema`

### LINKED ISSUE #25300: Active Record does not dump foreign key's referential actions with MySQL's ANSI_QUOTES
### Steps to reproduce

1. Prepare a schema in MySQL with a foreign key that has a referential action, e.g. `on delete cascade`.
1. Set your Active Record connection to use `sql_mode = ANSI_QUOTES`
1. Dump the schema using `rake db:schema:dump`

### Expected behavior
`db/schema.rb` should include `on_delete: :cascade` option in `add_foreign_key`.

### Actual behavior
`on_delete` option is missing. If you work without `sql_mode = ANSI_QUOTES` it is there.

### Diagnosis
Code to reproduce: https://gist.github.com/buehmann/55e155b4b56f75fe1a755bcab82f1a02

In the following line, Active Record tries to parse referential actions out of `show create table` output, but only expects ``` `this` ``` style of quoting, not `"that"` one.
https://github.com/rails/rails/blob/706f7e9c3e2a4807e37b1a1ef21febd6444adf6f/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L895

```
irb(main):017:0> @connection.execute("set session sql_mode = ''")
   (0.2ms)  set session sql_mode = '' /*application:Banking*/
=> nil
irb(main):018:0> @connection.select_one("SHOW CREATE TABLE a")["Create Table"]
   (0.3ms)  SHOW CREATE TABLE a /*application:Banking*/
=> "CREATE TABLE `a` (\n  `id` int(11) NOT NULL AUTO_INCREMENT,\n  `b_id` int(11) DEFAULT NULL,\n  PRIMARY KEY (`id`),\n  KEY `index_a_on_b_id` (`b_id`) USING BTREE,\n  CONSTRAINT `index_a_on_b_id` FOREIGN KEY (`b_id`) REFERENCES `b` (`id`) ON DELETE CASCADE\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci"
irb(main):019:0> @connection.execute("set session sql_mode = 'ansi_quotes'")
   (0.2ms)  set session sql_mode = 'ansi_quotes' /*application:Banking*/
=> nil
irb(main):020:0> @connection.select_one("SHOW CREATE TABLE a")["Create Table"]
   (0.3ms)  SHOW CREATE TABLE a /*application:Banking*/
=> "CREATE TABLE \"a\" (\n  \"id\" int(11) NOT NULL AUTO_INCREMENT,\n  \"b_id\" int(11) DEFAULT NULL,\n  PRIMARY KEY (\"id\"),\n  KEY \"index_a_on_b_id\" (\"b_id\") USING BTREE,\n  CONSTRAINT \"index_a_on_b_id\" FOREIGN KEY (\"b_id\") REFERENCES \"b\" (\"id\") ON DELETE CASCADE\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci"
```

### System configuration
**Rails version**: `Using rails 5.1.0.alpha from git://github.com/rails/rails.git (at master@3fc5577)`

**Ruby version**: `ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]`

**MySQL version**: `mysqld  Ver 5.6.27 for osx10.10 on x86_64 (Homebrew)`

### PR DESCRIPTION
**Pull request by kamipo**
Fixes #25300.

### CODE REVIEW

File: activerecord/test/schema/schema.rb [hunk 1]
```diff
@@ -409,6 +409,14 @@
     t.references :student
   end
 
+  create_table :students, force: true do |t|
+    t.string :name
+    t.boolean :active
+    t.integer :college_id
+  end
+
+  add_foreign_key :lessons_students, :students, on_delete: :cascade
```
Comment (sgrif) [inline activerecord/test/schema/schema.rb]: Why does this need to move?

Comment (kamipo) [inline activerecord/test/schema/schema.rb]: Currently `schema.rb` for tests have three `add_foreign_key`.

* `add_foreign_key :authors, :author_addresses`  
https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/schema/schema.rb#L55-L66

```ruby
  create_table :authors, force: true do |t|
    t.string :name, null: false
    t.integer :author_address_id
    t.integer :author_address_extra_id
    t.string :organization_id
    t.string :owned_essay_id
  end

  create_table :author_addresses, force: true do |t|
  end

  add_foreign_key :authors, :author_addresses
```

* `add_foreign_key :fk_test_has_fk, :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"`  
https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/schema/schema.rb#L997-L1004

```ruby
    create_table :fk_test_has_fk, force: true do |t|
      t.integer :fk_id, null: false
    end

    create_table :fk_test_has_pk, force: true, primary_key: "pk_id" do |t|
    end

    add_foreign_key :fk_test_has_fk, :fk_test_has_pk, column: "fk_id", name: "fk_name", primary_key: "pk_id"
```

* `add_foreign_key :lessons_students, :students`  
https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/schema/schema.rb#L407-L410  
https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/schema/schema.rb#L780-L784  
https://github.com/rails/rails/blob/v5.0.0.rc1/activerecord/test/schema/schema.rb#L1005

For a table that has a foreign key must be removed prior to the table that is referenced by a foreign key, I think it is important that the order of `create_table` that is dependent on foreign key is easy to understand.

**Comment by eileencodes** [comment]
Interestingly this fixes 2 bugs, one that I experienced. When you, @kamipo, added `create_table_info` caching in #21664 and #22547 it ended up caching the foreign key statement if you added an fk and then later altered your foreign key `on_update` or `on_delete` options. The result was the migration would run but the schema dump wouldn't get upgraded. This change fixes that bug because it no longer uses the regex on `create_table_info`. I'm going to merge this and add a regression test for the caching problem I saw in case this for some reason gets reverted (we don't want to reintroduce a caching error again).

**Comment by buehmann** [comment]
Thanks a lot for fixing this!

**Comment by rafaelfranca** [comment]
Backported in 72d18e5

**Pull request merged**